### DISCUSSION TITLE
연관관계 편의 메서드 작성 위치

### DISCUSSION THREAD

**Discussion by songsunkook** [BCSDLab/KOIN_API_V2#151] [에러 / 고민사항]
> 디스코드 채널에서 나온 질문을 옮긴 discussion입니다.

# 고민사항
`GET /dept`, `GET /detps` API 개발 도중 연관관계 편의 메서드를 작성하려고 하는데 어느 위치에 작성하는 것이 좋을지 고민중입니다.

## 현재 상황
현재 테이블 연관관계 구조는 아래와 같습니다.

`deptInfo(1) : deptNums(N) [일대다 양방향 관계]`
```
deptInfo는 List<DeptNum>을 가진다.
deptNum은 DeptInfo를 가진다.
```

찾아보니 연관관계 편의 메서드는 주인에 상관없이 많이 사용하는 쪽에 두면 된다고 하는데, 사용 빈도는 두 객체가 동일합니다.

### deptInfo에 둘 경우

- 연관관계 편의 메서드 위치에 따른 구조
  - List로써 deptNums를 유지하기 때문에 연관관계 편의 메서드 작성 시 `addDeptNum()`과 `removeDeptNum()` 메서드 두 가지를 작성해야 합니다.
  - 근거: 여러 deptNum을 유지할 수 있어야 하기 때문에 하나의 deptNum을 추가한다고 기존 deptNum을 지워야 하는 것은 아닙니다.
- 고민사항
  - deptNum을 List로 유지하기 때문에 builder 단계를 스킵하고 연관관계 편의 메서드 호출이 가능합니다.
  - deptInfo는 연관관계 편의 메서드에서 deptNum을 언급하고, deptNum은 builder에서 deptInfo를 언급합니다.
- 양측에서 서로를 언급하기에 양방향 연관관계로써 적절하다고 판단했습니다.

### deptNum에 둘 경우

- 연관관계 편의 메서드 위치에 따른 구조
  - 단일 deptInfo를 유지하기 때문에 연관관계 편의 메서드 작성 시 `setDeptInfo()`만 작성해도 됩니다.
  - 근거: 하나의 deptInfo를 유지하기 때문에 하나의 deptInfo를 설정하려면 기존 deptInfo를 지워야 합니다.
- 고민사항
  - deptInfo를 단일 객체로 유지하는데 필수 요소이기 때문에 builder 단계를 스킵할 수 없습니다.
    - builder 단계에서 deptInfo 선언 후 연관관계 편의 메서드 호출 시 재매핑되는 상황이 발생합니다.(확실하지 않음)
  - deptInfo는 deptNum을 언급하지 않고, deptNum은 builder와 연관관계 편의 메서드에서 한 번씩 총 두번 deptInfo를 언급합니다.
- 양측에서 서로를 언급하지 않고 한 쪽에서 두 번 언급하기에 적절하지 않다고 판단했습니다.

## 결론

저는 deptInfo에 연관관계 편의 메서드를 두는 것이 적절하다고 판단했는데, 혹시 다른 분들은 어떻게 생각하시는지 궁금합니다.

cc. @Choi-JJunho @Invidam @daheeParkk

**Comment by Invidam** [comment]
## 기존에 단 댓글
기타) DeptNum의 빌더 단계에서 DeptInfo를 넣지 않도록 수정할 수 있습니다. (생성자의 인자에서 DeptInfo를 제거)

**찾아보니 연관관계 편의 메서드는 주인에 상관없이 많이 사용하는 쪽에 두면 된다고 하는데** 에 대해 왜 그런지 생각해봐야 할 것 같습니다.

제 생각에는 `JPA-편의 메서드를 사용한다` =` 강건성이 떨어진다 `라고 생각합니다. (어디에 두든지 무관하게) 따라서 성능도 똑같고 강건성도 똑같으니 실용성(많이 사용하는 쪽에 두자)는 이야기를 하는 것 같아요.

어디에 두든 실용성이 같다고 했는데, 과연 그럴까요? 현재 몇 번 사용되는지만 고려할 게 아니라
- 다양한 학과 정보가 추가되어도 빈도가 같을지
- 학과 정보에 대한 CUD가 추가되어도 빈도가 같을지
생각해봐야 할 것 같습니다.

**Comment by Invidam** [comment]
## 지금 생각
23학번부터 학과정보를 `000`으로 통일한다고 알고 있습니다. 그렇다면 학과 정보 저장에 대한 논의(연관관계 메서드에 대한 논의)가 크게 의미가 없을 것 같아 **아무거나 빨리**하는 게 최선의 선택 같습니다.
